src/comparePayloads.js

```diff
@@ -0,0 +1,180 @@
+      repository: compareObjects("type", version.repository, comparedVersion.repository)
+        ?? compareObjects("url", version.repository, comparedVersion.repository),
+      scripts: compareDictionnaries(version.scripts, comparedVersion.scripts),
+      warnings: arrayDiff(version.warnings, comparedVersion.warnings),
+      composition: compareComposition(version.composition, comparedVersion.composition),
+      licenseIds: arrayDiff(version.license.uniqueLicenseIds, comparedVersion.license.uniqueLicenseIds),
+      flags: arrayDiff(version.flags, comparedVersion.flags),
+      links: compareValues(version.links, comparedVersion.links)
+    };
+
+    comparedVersions.set(name, diff);
+  }
+
+  return {
+    compared: comparedVersions,
+    ...versions
+  };
+}
+
+function compareComposition(original = {}, toCompare = {}) {
+  return {
+    minified: arrayDiff(original.minified, toCompare.minified),
+    required_thirdparty: arrayDiff(original.required_thirdparty, toCompare.required_thirdparty),
+    required_nodejs: arrayDiff(original.required_nodejs, toCompare.required_nodejs),
+    required_unused: arrayDiff(original.required_unused, toCompare.required_unused),
+    required_missing: arrayDiff(original.required_missing, toCompare.required_missing)
+  };
+}
+
+function compareDictionnaries(original, toCompare) {
+  const { comparable, ...diff } = dictionariesDiff(original, toCompare);
+
+  const compared = new Map();
+  for (const [name, [entity, comparedEntity]] of comparable) {
+    compared.set(name, compareValues(entity, comparedEntity));
+  }
+
+  return {
+    compared,
+    ...diff
+  };
+}
+
+function compareObjects(key, original = {}, toCompare = {}) {
+  if (original[key] === toCompare[key]) {
+    return undefined;
+  }
+
+  return {
+    prev: original,
+    now: toCompare
+  };
+}
+
+function compareValues(original, toCompare) {
+  if (typeof original === "object") {
+    if (JSON.stringify(original) === JSON.stringify(toCompare)) {
+      return undefined;
+    }
+  }
+  else if (original === toCompare) {
+    return undefined;
+  }
+
+  return {
+    prev: original,
+    now: toCompare
+  };
+}
+
+function dictionariesDiff(original = {}, toCompare = {}) {
+  const added = new Map();
+  const removed = new Map();
+  const comparable = new Map();
+
+  Object.keys(original).forEach((key) => {
+    if (Object.hasOwn(toCompare, key)) {
+      comparable.set(key, [original[key], toCompare[key]]);
+    }
+    else {
+      removed.set(key, original[key]);
+    }
+  });
+
+  Object.keys(toCompare).forEach((key) => {
+    if (!Object.hasOwn(original, key)) {
+      added.set(key, toCompare[key]);
+    }
+  });
+
+  return { added, removed, comparable };
+}
+
+function arrayDiff(original = [], toCompare = []) {
+  const added = toCompare.filter((v, i) => {
+    if (typeof v !== "object") {
+      return v !== original[i];
+    }
+
+    return JSON.stringify(v) !== JSON.stringify(original[i]);
```

Comment:
I'm not a big fan of `JSON.stringify`, it's slow but it's probably fine
Maybe we can compare `original.length` & `toCompare.length` in first step ? It could slightly improve performances, wdyt ?

Also, `arrayDiff` is not very explicit; `arrayEquals` would be clearer, wdyt ?

Reply:
I think it's a good idea, also there is `fast-json-stringify` but I tried to not add new dependencies...

Reply:
> I'm not a big fan of `JSON.stringify`, it's slow but it's probably fine Maybe we can compare `original.length` & `toCompare.length` in first step ? It could slightly improve performances, wdyt ?

You mean this first check?
```js
if (v.length !== original[i].lengh) {
  return false;
}
```

> Also, `arrayDiff` is not very explicit; `arrayEquals` would be clearer, wdyt ?

There are two types of functions: **diff** functions and **comparing** functions. **diff** functions return an object with `added` and `removed` props (eventually `comparable`). It is not really about equality check but to know if entities have been added, removed and eventually to know what are comparable in case we want to compare them, tough it gives information about equality.

Reply:
Yeah my bad, with removed and added we can't check length and the function name is fine.

Reply:
Despite its limitations, `JSON.stringify` + string comparison is much faster than plain object key-by-key comparison.  If we remain consistent with the order of properties of objects in payloads, this shouldn't really be a problem, I think.